File: src/brs/web/api/http/handler/ParseTransaction.java
package brs.web.api.http.handler;

import brs.BurstException;
import brs.Transaction;
import brs.services.ParameterService;
import brs.services.TransactionService;
import brs.util.Convert;
import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.JSONData;
import brs.web.api.http.common.LegacyDocTag;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.Parameters.TRANSACTION_BYTES_PARAMETER;
import static brs.web.api.http.common.Parameters.TRANSACTION_JSON_PARAMETER;
import static brs.web.api.http.common.ResultFields.*;

public final class ParseTransaction extends ApiServlet.JsonRequestHandler {

  private static final Logger logger = LoggerFactory.getLogger(ParseTransaction.class);

  private final ParameterService parameterService;
  private final TransactionService transactionService;

  public ParseTransaction(ParameterService parameterService, TransactionService transactionService) {
    super(new LegacyDocTag[] {LegacyDocTag.TRANSACTIONS}, TRANSACTION_BYTES_PARAMETER, TRANSACTION_JSON_PARAMETER);
    this.parameterService = parameterService;
    this.transactionService = transactionService;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) throws BurstException {

    String transactionBytes = Convert.emptyToNull(req.getParameter(TRANSACTION_BYTES_PARAMETER));
    String transactionJSON = Convert.emptyToNull(req.getParameter(TRANSACTION_JSON_PARAMETER));
    Transaction transaction = parameterService.parseTransaction(transactionBytes, transactionJSON);
    JsonObject response = JSONData.unconfirmedTransaction(transaction);
    try {
      transactionService.validate(transaction);
    } catch (BurstException.ValidationException|RuntimeException e) {
      logger.debug(e.getMessage(), e);
      response.addProperty(VALIDATE_RESPONSE, false);
      response.addProperty(ERROR_CODE_RESPONSE, 4);
      response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid transaction: " + e.toString());
      response.addProperty(ERROR_RESPONSE, e.getMessage());
    }
    response.addProperty(VERIFY_RESPONSE, transaction.verifySignature() && transactionService.verifyPublicKey(transaction));
    return response;
  }

}

File: src/main/java/brs/http/ParseTransaction.kt
package brs.http

import brs.BurstException
import brs.services.ParameterService
import brs.services.TransactionService
import com.google.gson.JsonElement
import org.slf4j.LoggerFactory

import javax.servlet.http.HttpServletRequest

import brs.http.common.Parameters.TRANSACTION_BYTES_PARAMETER
import brs.http.common.Parameters.TRANSACTION_JSON_PARAMETER
import brs.http.common.ResultFields.ERROR_CODE_RESPONSE
import brs.http.common.ResultFields.ERROR_DESCRIPTION_RESPONSE
import brs.http.common.ResultFields.ERROR_RESPONSE
import brs.http.common.ResultFields.VALIDATE_RESPONSE
import brs.http.common.ResultFields.VERIFY_RESPONSE
import brs.util.convert.emptyToNull

internal class ParseTransaction(private val parameterService: ParameterService, private val transactionService: TransactionService) : APIServlet.JsonRequestHandler(arrayOf(APITag.TRANSACTIONS), TRANSACTION_BYTES_PARAMETER, TRANSACTION_JSON_PARAMETER) {

    override suspend fun processRequest(request: HttpServletRequest): JsonElement {

        val transactionBytes = request.getParameter(TRANSACTION_BYTES_PARAMETER).emptyToNull()
        val transactionJSON = request.getParameter(TRANSACTION_JSON_PARAMETER).emptyToNull()
        val transaction = parameterService.parseTransaction(transactionBytes, transactionJSON)
        val response = JSONData.unconfirmedTransaction(transaction)
        try {
            transactionService.validate(transaction)
        } catch (e: BurstException.ValidationException) {
            logger.debug(e.message, e)
            response.addProperty(VALIDATE_RESPONSE, false)
            response.addProperty(ERROR_CODE_RESPONSE, 4)
            response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid transaction: $e")
            response.addProperty(ERROR_RESPONSE, e.message)
        } catch (e: RuntimeException) {
            logger.debug(e.message, e)
            response.addProperty(VALIDATE_RESPONSE, false)
            response.addProperty(ERROR_CODE_RESPONSE, 4)
            response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid transaction: $e")
            response.addProperty(ERROR_RESPONSE, e.message)
        }

        response.addProperty(VERIFY_RESPONSE, transaction.verifySignature() && transactionService.verifyPublicKey(transaction))
        return response
    }

    companion object {

        private val logger = LoggerFactory.getLogger(ParseTransaction::class.java)
    }

}
